### ISSUE TITLE
Avoid translating internal routers ip address to localhost when pinging

### ISSUE THREAD

**Issue by hfakoor222** [GNS3/gns3-server#2227]
I'm pinging internally from a bridged topology to my loopback, when I'm running a socket script against the ping I'm seeing the source as my loopbaack and dest as loopback. The host pc is translating the routers address to localhost. is there a way to avoid this?

**Comment by josephmhiggins** [comment]
gns3 version?
operating system name and version?
hypervisor name and version?

**Comment by hfakoor222** [comment]
GNS:  2,2,38
Hypervisor:  VmWare Workstation Pro, I just downloaded it this month,  so version 17
OS: Windows 10 Home 22H2

I also noticed today internal routing wasn't on so i changed the registries to allow it, not sure if it will make a difference.

The translation happens in NAT and bridged mode. I was assuming at least in bridge the python script should've been able to strip the header properly.

Using wireshark inside the GNS evironment shows the correct source.

I'm open to any suggestions, not the end of the world, as I may be moving to Linux, but it would be nice if I found a solution - this could very well be and is probably a Windows issue on my end.

**Comment by josephmhiggins** [comment]
Why are you getting involved with a windows loopback interface?


The gns3 vm cloud to vmware vmnet interface translation should be like so

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/joe_I7/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/joe_I7/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72" lang=EN-US style='tab-interval:.5in;
word-wrap:break-word'>



Host adapters | GNS3 VM adapters | GNS3 Cloud interfaces
-- | -- | --
vmnet0 Bridged | eth2 | eth2
vmnet1 Host-Only | eth0 | eth0
vmnet8 NAT | eth1 | eth1
vmnet2 Custom Host-Only | eth3 | eth3



</body>

</html>

This word document attach provides instructions to properly configure vmware workstation pro, in this case 16.2.5, on windows 10.

[configuring vmware workstation pro for the gns3 vm.docx](https://github.com/GNS3/gns3-server/files/11714635/configuring.vmware.workstation.pro.for.the.gns3.vm.docx)

**Comment by josephmhiggins** [comment]
The docx can be replaced by code.
VMware supports a rest api.
The entire properly configured vmware network editing can be done via the rest api.
Your situation is:
I want to accomplish A
and I will accomplish A by doing B

Your B is False.
What is your A?

**Comment by josephmhiggins** [comment]
If you look at the gns3 gui, then you will see there is a missing Dock.
The main problem is that VMware supports 100 vmnets (or 200 vmnets - I forget which)

The dock would look like this on my Windows 10 host

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/joe_I7/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/joe_I7/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72" lang=EN-US style='tab-interval:.5in;
word-wrap:break-word'>



Host adapters | GNS3 VM links     (found via Ubuntu ip link command) | GNS3 Cloud interfaces | Status
-- | -- | -- | --
vmnet0 Bridged | eth2 | eth2 | Green
vmnet1 Host-Only | eth0 | eth0 | Green
vmnet8 NAT | eth1 | eth1 | Green
vmnet2 Custom Host-Only | eth3 | eth3 | Green



</body>

</html>

**Comment by josephmhiggins** [comment]
brilliant. all GNS3 VM links that connect to vmware vmnets HAVE ipv6 link-local addresses
so that is it.

ping -6 -t fe80::20c:29ff:fe7a:1b7e
will provoke GNS3 VM eth 2, (connected my vmnet0 bridged interface)
from showing up where it belongs, to be my neighbor on
Vlan0033-SW241-G5

ADMIN > Get-NetNeighbor -IPAddress fe80::20c:29ff:fe7a:1b7e | Where-Object {$_.State -eq 'Reachable'}
ifIndex IPAddress                                          LinkLayerAddress      State       PolicyStore    
------- ---------                                          ----------------      -----       -----------    
17      fe80::20c:29ff:fe7a:1b7e                           00-0C-29-7A-1B-7E     Reachable   ActiveStore    

ADMIN > get-netadapter -InterfaceIndex 17
Name                      InterfaceDescription                    ifIndex Status       MacAddress             LinkSpeed
----                      --------------------                    ------- ------       ----------             ---------
Vlan0033-SW241-G5         Intel(R) 82576 Gigabit Dual Port N...#5      17 Up           00-1B-21-38-4A-05         1 Gbps

In other words, all interfaces will show up as Green by ping the link-local ipv6 address.

The only way this may be able to stop this is if a gns3 user disables ipv6.
....But they would need to know how to do that - NOT likely for a new gns3 user; it is easily detectable if they do.

getting all the ipv6 link-local on the GNS3 VM  might take me about 40hours to figure out how to do 👎

**Comment by hfakoor222** [comment]
I'm going to run some more testing. The order of adapter settings helped. I'm listening in on a loopback of my local PC as thats where the ping is being sent to trigger Python. I could just write the script to listen inside the GNS environment without translations happening, but the production code will be listening in on a loopback, to trigger network monitoring of devices based off of pings after router events happen. I will try it and post results at some point. The order of adapter operations helped as mentioned in the docx,  thank you.

**Comment by josephmhiggins** [comment]
From almost the time GNS3 started, people have been using the loopback interface to provide connectivity the gns3 topology and the internet.

This was a catastrophic blunder.
The developers fixed this with vmware bridged interface vmnet0
The loopback interface was also used by folks with only a wi-fi adapter.
This was another catastrophic blunder.
I fixed that with vmnet2.

vmnet0 is the gns3 developer's vmnet for the GNS3 VM.
vmnet1 is the gns3 developer's vmnet for the GNS3 Nat appliance.

I obviously do not know anything about your "production" code.
But if your production code is to support gns3 topologies - than your design needs to be reevaluated.
VMware only supports bridging vmnet interfaces to physical devices - per their documentation.
A loopback interface is a virtual interface.
Strictly by the book, whether this is true or not, I say vmware's does not support bridging a vmnet to a loopback and then bridging the loopback to an internet facing LAN.

....of course, this could all change in gns3 version 3.0

**Comment by josephmhiggins** [comment]
here's the code in vmware.py
https://github.com/GNS3/gns3-server/tree/master/gns3server/compute/vmware

You can read python.
I can not read python.

**Comment by hfakoor222** [comment]
> From almost the time GNS3 started, people have been using the loopback interface to provide connectivity the gns3 topology and the internet.
> 
> This was a catastrophic blunder. The developers fixed this with vmware bridged interface vmnet0 The loopback interface was also used by folks with only a wi-fi adapter. This was another catastrophic blunder. I fixed that with vmnet2.
> 
> vmnet0 is the gns3 developer's vmnet for the GNS3 VM. vmnet1 is the gns3 developer's vmnet for the GNS3 Nat appliance.
> 
> I obviously do not know anything about your "production" code. But if your production code is to support gns3 topologies - than your design needs to be reevaluated. VMware only supports bridging vmnet interfaces to physical devices - per their documentation. A loopback interface is a virtual interface. Strictly by the book, whether this is true or not, I say vmware's does not support bridging a vmnet to a loopback and then bridging the loopback to an internet facing LAN.
> 
> ....of course, this could all change in gns3 version 3.0

Hey sorry, just checked gmail.  Thanks for the reply. The production code doesn't rely on GNS, I needed a emulated router environment,to send eem pings, strip headers and perform actions based on those pings. Now EEM pings respond to many events, and I wrote a 24/7 listener to run scripts based on icmp data patterns or number of pings etc etc. What was killing me was the headers were stripping wrong source addresses, a somewhat common problem I saw with C socket program issues on stack overflow. I decided to turn routing on for my Windows 10, did a reboot and haven't had a chance to check if that solved it. I could run the code on any interface other than loopback. The info you've given has been super informative.  I'm going to reread the docx (I fast read it once, and noticed the order of operations for the adapters, which may fix my issue without even turning 'routing' on inside my Windows 10 registry...), and I wasn't aware of the loopback issue before this.  I may just play around with everything and see what happens.

**Comment by hfakoor222** [comment]
Also with or without GNS, I had wrong source addresses from sending pings from the CLI. This told me something inside my PC is translating the address - I'm not sure if the listening implementation I've set up strips reply to pings or incoming pings. Maybe the 'turn routing on' solved this. I got the listener from chapter 3 Black hat python, and looking at other examples it worked for them.

**Comment by hfakoor222** [comment]
The virtual loopback interface issue could very well be the issue - but I believe i disticntly remember testing this on multiple interfaces with the same result. Again leading me to think it's a Windows environment issue.

**Comment by hfakoor222** [comment]
If the pings are coming for an outside host I don't expect translation to be a problem, in a productions setting.

**Comment by josephmhiggins** [comment]
My docx is  nice. But if you can read pthon - the gns3 developers code is better.
However, the gns3 code supports vmware version 11, so some aspects  the code is no longer valid.
If you are talking about Cisco EEM, then the best in the world is Joseph Doherty over at cisco.com; he's wonderful.

**Comment by josephmhiggins** [comment]
Windows 10 has some bugs in their code.
The windows 10 gns3server will try to reach the gns3 vm on the "internet" interface when I shutdown the gns3 vm.
This request hits my firewall and sends a log to my kiwi.
I know about it and have left the problem alone.
The GNS3 VM is on vmware vmnet1 192.168.161.0/24.
Windows (the gns3 server) tries to reach the GNS3 VM on my internet interface, 192.168.33.0/24.

Normal Windows GNS3 users do not understand Windows firewall(wf.msc).
So, I never tell them to change anything and to just leave the windows firewall alone.
But strictly speaking, by the book, the proper procedure would be to create outbound entries in the firewall - turn on logging to pfirewall.log; change the ACL entry for pfirewall log - fire up a detector to a sylog server for any changes in pfirewall.log - to check that no "reverse path forwarding" packets were being generated by Windows 10.

You would need a "poller" not an asynchronous "watch-onchange for that pfirewall.log.
But yeah, I would definitely tinker with the windows firewall and close down traffic in an outbound fashion for troubleshooting purposes.

That is the proper procdure.

..........again.....this might change in gns3 version 3.0 because gns3 version3.0 will have more security and should be designed to not accept "hello" messages destined for "eth0" on its "NAT" or "bridged" interface
